### PR TITLE
FOGL-1487 - Create a addScheduledTask REST entry point that is similar to addService

### PR DESCRIPTION
**Pull request by amarendra-dianomic**
Steps to test:
1. make clean, make, scripts/foglamp reset
2. Start server.
3. Check scripts/foglamp status + syslog that NO 'HTTP North' process is running.
4. In another folder, download, extract and run
[ldeploy-http-north-plugin.sh.zip](https://github.com/foglamp/FogLAMP/files/2191759/ldeploy-http-north-plugin.sh.zip). Before executing this script, please ensure that foglamp-north-http/cmds-http-north.sql contains only below commands:
```
-- SQLite only
-- Plugin loader configuration
INSERT INTO destinations ( id, description)
    VALUES ( 2, 'HTTP North' );

INSERT INTO streams ( id, destination_id, description, last_object )
    VALUES ( 3, 2, 'HTTP North', 0 );

-- Statistics
INSERT INTO statistics (key, description, value, previous_value)
    VALUES('SENT_3', 'Readings data sent via HTTP North', 0, 0);

```
5. Check scripts/foglamp status + syslog that now 'HTTP North' process is running.
6. We may also check the actual sending process by ingesting data via, say fogbench, and see if the sending process via HTTP North is actually working. 

Test result:
`=================================================== 1158 passed, 31 skipped in 84.67 seconds ====================================================
`

### CODE REVIEW

File: python/foglamp/services/core/api/task.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+     curl -X POST http://localhost:8081/foglamp/scheduled/task -d 
+     '{
+        "name": "North Readings to PI",
+        "plugin": "omf",
+        "enabled": true,
+        "type": 3,
+        "day": 0,
+        "time": 0,
+        "repeat": 30,
+        "with_configuration": {
```
[the remark this answers is not in the document]
Comment (amarendra-dianomic) [reply python/foglamp/services/core/api/task.py]: Please see https://scaledb.atlassian.net/browse/FOGL-1487?focusedCommentId=39328&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-39328

**Comment by amarendra-dianomic** [comment]
@MarkRiddoch @praveen-garg feedback changes done. Please check.

File: python/foglamp/services/core/api/task.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+            raise web.HTTPBadRequest(reason='Missing name property in payload.')
+        if plugin is None:
+            raise web.HTTPBadRequest(reason='Missing plugin property in payload.')
+        if task_type is None:
+            raise web.HTTPBadRequest(reason='Missing type property in payload.')
+        if utils.check_reserved(name) is False:
+            raise web.HTTPBadRequest(reason='Invalid name property in payload.')
+        if utils.check_reserved(plugin) is False:
+            raise web.HTTPBadRequest(reason='Invalid plugin property in payload.')
+        if task_type not in ['south', 'north']:
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/task.py]: keep it a list but only with north, hence `['north']`

File: python/foglamp/services/core/api/task.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+            raise web.HTTPBadRequest(reason='Only north and south types are supported.')
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/task.py]: `Only north type is supported.`  

Or better

```
supported_task_type = ['north']
if task_type not in supported_task_type:
       raise web.HTTPBadRequest(reason='Only {} types are supported.'.format(', '.join(supported_task_type))) 
```

File: python/foglamp/services/core/api/task.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+# See: http://foglamp.readthedocs.io/
+# FOGLAMP_END
+
+import datetime
+from aiohttp import web
+from foglamp.common.storage_client.payload_builder import PayloadBuilder
+from foglamp.common.configuration_manager import ConfigurationManager
+from foglamp.services.core import server
+from foglamp.services.core import connect
+from foglamp.services.core.scheduler.entities import Schedule, StartUpSchedule, TimedSchedule, \
```
[the remark this answers is not in the document]
Comment (praveen-garg) [reply python/foglamp/services/core/api/task.py]: @amarendra-dianomic  please remove this specific import

File: python/foglamp/services/core/api/task.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+_help = """
+    -------------------------------------------------------------------------------
+    | GET POST            | /foglamp/scheduled/task                               |
+    -------------------------------------------------------------------------------
+"""
+
+
+#################################
+#  Task
+#################################
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/task.py]: remove this please

File: python/foglamp/services/core/api/task.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+        if task_type not in ['south', 'north']:
+            raise web.HTTPBadRequest(reason='Only north and south types are supported.')
+        if cmd_params is not None:
+            if not isinstance(cmd_params, dict):
+                raise web.HTTPBadRequest(reason='cmd_params must be a dict.')
+
+        if schedule_type is None:
+            raise web.HTTPBadRequest(reason='Schedule Type is mandatory')
+        if not isinstance(schedule_type, int) and not schedule_type.isdigit():
+            raise web.HTTPBadRequest(reason='Error in type: {}'.format(schedule_type))
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/task.py]: Schedule type error OR `Error in schedule type` as in payload type is for task type

File: python/foglamp/services/core/api/task.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+        if int(schedule_type) not in list(Schedule.Type):
+            raise web.HTTPBadRequest(reason='Schedule type error: {}'.format(schedule_type))
+        if int(schedule_type) == Schedule.Type.STARTUP:
+            raise web.HTTPBadRequest(reason='Schedule type cannot be STARTUP: {}'.format(schedule_type))
+
+        schedule_type = int(schedule_type)
+
+        if schedule_day is not None:
+            if isinstance(schedule_day, float) or (isinstance(schedule_day, str) and (schedule_day.strip() != "" and not schedule_day.isdigit())):
+                raise web.HTTPBadRequest(reason='Error in day: {}'.format(schedule_day))
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/task.py]: `schedule_day must be a valid integer, 1 (Monday) to 7 (Sunday)`

File: python/foglamp/services/core/api/task.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+        else:
+            schedule_day = int(schedule_day) if schedule_day is not None else None
+
+        if schedule_time is not None and (not isinstance(schedule_time, int) and not schedule_time.isdigit()):
+            raise web.HTTPBadRequest(reason='Error in time: {}'.format(schedule_time))
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/task.py]: return  the error with required format info and actual payload property name i.e. schedule_time

File: python/foglamp/services/core/api/task.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+        else:
+            schedule_time = int(schedule_time) if schedule_time is not None else None
+
+        if schedule_repeat is not None and (not isinstance(schedule_repeat, int) and not schedule_repeat.isdigit()):
+            raise web.HTTPBadRequest(reason='Error in repeat: {}'.format(schedule_repeat))
+        else:
+            schedule_repeat = int(schedule_repeat) if schedule_repeat is not None else None
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/task.py]: as above

File: python/foglamp/services/core/api/task.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+
+        if schedule_type == Schedule.Type.TIMED:
+            if not schedule_time:
+                raise web.HTTPBadRequest(reason='Schedule time cannot be empty/None for TIMED schedule.')
+            if schedule_day is not None and (not (schedule_day < 1 or schedule_day > 7)):
+                raise web.HTTPBadRequest(reason='Day must either be None or must be an integer and in range 1-7.')
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/task.py]: you got it. it should be used the same way,  above ^

File: python/foglamp/services/core/api/task.py [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+            if schedule_time < 0 or schedule_time > 86399:
+                raise web.HTTPBadRequest(reason='Time must be an integer and in range 0-86399.')
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/task.py]: yeah!

File: python/foglamp/services/core/api/task.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+
+        if schedule_type == Schedule.Type.INTERVAL:
+            if schedule_repeat is None:
+                raise web.HTTPBadRequest(reason='Repeat is required for INTERVAL Schedule type.')
+            elif not isinstance(schedule_repeat, int):
+                raise web.HTTPBadRequest(reason='Repeat must be an integer.')
```
[the remark this answers is not in the document]
Comment (praveen-garg) [reply python/foglamp/services/core/api/task.py]: must be an integer and expected input format, range or value?

File: python/foglamp/services/core/api/task.py [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+            m, s = divmod(schedule_time if schedule_time is not None else 0, 60)
+            h, m = divmod(m, 60)
+            schedule.time = datetime.time().replace(hour=h, minute=m, second=s)
+            schedule.repeat = datetime.timedelta(schedule_repeat if schedule_repeat is not None else 0)
+            schedule.exclusive = True
+            schedule.enabled = False  # if "enabled" is supplied, it gets activated in save_schedule() via is_enabled flag
+
+            # Save schedule
+            await server.Server.scheduler.save_schedule(schedule, is_enabled)
+            schedule = await server.Server.scheduler.get_schedule_by_name(name)
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/task.py]: does not `save_schedule` return the schedule_id ?

File: python/foglamp/services/core/api/task.py [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+            raise web.HTTPInternalServerError(reason='Failed to created schedule. {}'.format(ex.error))
+        except Exception as ins_ex:
+            await revert_configuration(storage, name)  # Revert configuration entry
+            await revert_scheduled_processes(storage, name)  # Revert scheduled_process entry
+            raise web.HTTPInternalServerError(reason='Failed to created schedule. {}'.format(str(ins_ex)))
+
+        return web.json_response({'name': name, 'id': str(schedule.schedule_id)})
+
+    except ValueError as ex:
+        raise web.HTTPNotFound(reason=str(ex))
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/task.py]: what is not found?

File: python/foglamp/services/core/api/task.py (hunk 13 above)
Comment (amarendra-dianomic) [reply python/foglamp/services/core/api/task.py]: Nope

File: python/foglamp/services/core/api/task.py (hunk 14 above)
Comment (amarendra-dianomic) [reply python/foglamp/services/core/api/task.py]: changed to HTTPInternalServerError

**Comment by amarendra-dianomic** [comment]
@praveen-garg feedback changes done. Pl check.

**Comment by praveen-garg** [comment]
@amarendra-dianomic  Kindly update PR desc for http-north plugin. OMF and OCS plugins are  there OOB so adding http-north plugin example will be the right thing to do.

**Pull request merged**